### PR TITLE
Adds tox, Travis support and some tests

### PR DESCRIPTION
**Pull request by TimSimpsonR**

### CODE REVIEW

File: tests/test_mutualauth.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        return self.__role
+
+
+class FakeResource(object):
+
+    def __init__(self, authorized_for):
+        self.authorized_for = authorized_for
+
+
+class TestAuthorize(unittest.TestCase):
```
Comment (derwolfe) [inline tests/test_mutualauth.py:28]: Could these tests be written similarly to https://github.com/RackSec/overcorps/blob/master/tests/functional_tests.py#L439-L484? The reason being, falcon-mutualauth is being used specifically with falcon -  as such it seems like exercising that logic might be helpful.

Comment (derwolfe) [inline tests/test_mutualauth.py:28]: The other benefit of having tests similar to those in overcorps, is that these tests exercise the client certificate logic.

File: tox.ini [hunk 2]
```diff
@@ -0,0 +1,17 @@
+[tox]
+envlist = py27,py35,pep8
```
Comment (ehashman) [inline tox.ini]: Travis matrix has py34 as well but that appears to be missing here. I think it would be good to test---can we add it here?

File: tests/test_mutualauth.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    def _create_fake_resource(self):
+        return None
+
+    def test_log_in_as_admin(self):
+        self._auth.process_resource(FakeRequest('admin'), None, self._resource)
+
+    def test_log_in_as_principle(self):
+        self._auth.process_resource(
+            FakeRequest('principle'), None, self._resource)
```
Comment (ehashman) [inline tests/test_mutualauth.py:42]: I see that https://github.com/RackSec/falcon-mutualauth/blob/master/falcon_mutualauth/mutualauth.py#L78 splits on a comma---maybe we could exercise the set path by using multiple roles in one of these tests?

File: tests/test_mutualauth.py (hunk 1 above)
Comment (ehashman) [inline tests/test_mutualauth.py:28]: +1 to @derwolfe. However, I'm fine with saying this is out of scope for now and seeing it in another PR, fwiw.

File: .travis.yml [hunk 4]
```diff
@@ -0,0 +1,22 @@
+sudo: false
+language: python
+env:
+  global:
+    - CODECOV_TOKEN=003330f1-4898-4761-b1ea-555084097df0
```
Comment (derwolfe) [inline .travis.yml]: I believe, since this is an open source repo, you don't need to include the token. Per their example project: https://github.com/codecov/example-python/blob/master/.travis.yml

File: test-requirements-pep8.txt [hunk 5]
```diff
@@ -0,0 +1,2 @@
+flake8
+pep8-naming
```
Comment (derwolfe) [inline test-requirements-pep8.txt:2]: Are we using `pep8-naming` in our other projects?

Comment (derwolfe) [inline test-requirements-pep8.txt:2]: It looks like it is in aperture, account-service, rms-util, and rms-toolbox. (and giskard).

Comment (derwolfe) [inline test-requirements-pep8.txt:2]: We should probably throw in `flake8-import-order` for consistency as well.

File: tests/test_mutualauth.py (hunk 1 above)
Comment (fboxwala) [inline tests/test_mutualauth.py:28]: I was actually in the process of writing a similar test (similar to the overcorps one)

File: tox.ini (hunk 2 above)
Comment (TimSimpsonR) [inline tox.ini]: Oh weak, I didn't realize Travis wasn't honoring what was in Tox. I'd personally rather stick to only one version of Python 3 (matching whatever we do in prod) but this should include 3.4 so no one runs locally and then gets surprised by a Travis failure.

File: tests/test_mutualauth.py (hunk 1 above)
Comment (TimSimpsonR) [inline tests/test_mutualauth.py:28]: Sounds good, but let's do that but in another PR.

File: tox.ini (hunk 2 above)
Comment (reaperhulk) [inline tox.ini]: The envlist is just the set of environments that are run when you just do `tox` without a `TOXENV` env var and without a `-e` argument. Since we run tox with `TOXENV` to target specific environments (which allows us to parallelize the tests somewhat) travis effectively ignores this.

File: tests/test_mutualauth.py (hunk 1 above)
Comment (derwolfe) [inline tests/test_mutualauth.py:28]: Just to make sure I'm understanding - you'd prefer to test `MutualAuthRequest` in another PR and cover the `Authorize` class with these tests?

Comment (TimSimpsonR) [inline tests/test_mutualauth.py:28]: Correct.

**Comment by ehashman** [comment]
Coverage appears to be busted, which is causing Travis to report failures: https://travis-ci.org/RackSec/falcon-mutualauth/jobs/136807984#L209 :(

File: tests/test_mutualauth.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import falcon
+from falcon_mutualauth.mutualauth import Authorize
+
+
+class FakeRequest(object):
+
+    def __init__(self, role):
+        self.__role = role
+        self.env = {'REMOTE_ADDR': 'http://test.test'}
+        self.user_agent = 'pytest'
```
Comment (derwolfe) [inline tests/test_mutualauth.py:15]: Are `env` or `user_agent` used?

File: tests/test_mutualauth.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+import unittest
+
+import falcon
+from falcon_mutualauth.mutualauth import Authorize
+
+
+class FakeRequest(object):
+
+    def __init__(self, role):
+        self.__role = role
```
Comment (derwolfe) [inline tests/test_mutualauth.py]: Is this meant to signal that `role` should be a private instance variable? If so, a single underscore is conventional :).

File: tests/test_mutualauth.py (hunk 6 above)
Comment (derwolfe) [inline tests/test_mutualauth.py:15]: Answered on own inspection - yes, they are :).

File: tests/test_mutualauth.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+    def setUp(self):
+        self._auth = Authorize('admin', 'principle')
+        self._resource = self._create_fake_resource()
+
+    def _create_fake_resource(self):
+        return None
+
+    def test_log_in_as_admin(self):
+        self._auth.process_resource(FakeRequest('admin'), None, self._resource)
```
Comment (derwolfe) [inline tests/test_mutualauth.py:38]: It seems like all uses of `self._resource` could be replace with `None`.

File: .travis.yml [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  - python: 2.7
+    env: TOXENV=py27
+  - python: 3.4
+    env: TOXENV=py34
+  - python: 3.5
+    env: TOXENV=py35
+install: pip install tox codecov
+script:
+  - tox
+  - coverage run tests.py
```
Comment (derwolfe) [inline .travis.yml]: You can remove the `coverage run tests.py` call as tox will take care of running [this](https://github.com/RackSec/falcon-mutualauth/pull/4/files#diff-b91f3d5bd63fcd17221b267e851608e8R8)

File: tests/test_mutualauth.py (hunk 8 above)
Comment (TimSimpsonR) [inline tests/test_mutualauth.py:38]: If it's done this way, then the FakeResource will be used when this method runs in the subclass instead of None.

File: .travis.yml (hunk 9 above)
Comment (ehashman) [inline .travis.yml]: Overcorps pip installs `codecov` and runs that, as opposed to `coverage`. Now that I'm looking, this might be a typo. Per Chris, to automatically runs coverage.

The other thing we probably want here is to move the coverage checks into `after_success`. Otherwise, codecov will always run, even if the tests failed.

File: tests/test_mutualauth.py (hunk 3 above)
Comment (TimSimpsonR) [inline tests/test_mutualauth.py:42]: Great idea. I added a test for that.

File: tests/test_mutualauth.py (hunk 7 above)
Comment (TimSimpsonR) [inline tests/test_mutualauth.py]: Actually, private vars use double underscores. 

On that note, I am guessing the convention on the team is to stick to private vars. However, I've personally gotten into a habit of only using protected variables as private ones are annoying to access if you ever need to. Ironically this ended up being a rule in Effective Python (https://hacktec.gitbooks.io/effective-python/content/en/Chapter3/item27.html) so I'd be in favor of using protected here instead.

**Comment by TimSimpsonR** [comment]
@derwolfe So the import order plugin works now in Travis, but not on my box.  :(

File: tests/test_mutualauth.py (hunk 7 above)
Comment (derwolfe) [inline tests/test_mutualauth.py]: Per [Pep-8](https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles):

```
_single_leading_underscore : weak "internal use" indicator. E.g. from M import * does not import objects whose name starts with an underscore.

single_trailing_underscore_ : used by convention to avoid conflicts with Python keyword, e.g.

Tkinter.Toplevel(master, class_='ClassName')
__double_leading_underscore : when naming a class attribute, invokes name mangling (inside class FooBar, __boo becomes _FooBar__boo ; see below).

__double_leading_and_trailing_underscore__ : "magic" objects or attributes that live in user-controlled namespaces. E.g. __init__ , __import__ or __file__ . Never invent such names; only use them as documented.****
```

The convention I've always seen in python libraries is that a single leading underscore implies a private variable or function. :).

**Comment by derwolfe** [comment]
regarding tox and import order - this was resolved by running `tox -r`, correct?

**Comment by TimSimpsonR** [comment]
@derwolfe Yes. Just changed the double underscore to single too.

File: .travis.yml [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+  - python: 2.7
+    env: TOXENV=py27
+  - python: 3.4
+    env: TOXENV=py34
+  - python: 3.5
+    env: TOXENV=py35
+install: pip install tox codecov
+script:
+  - tox
+  - coverage run -m pytest
```
Comment (derwolfe) [inline .travis.yml]: The comment looks to have been lost, but this call to coverage should be removed. Tox takes care of this :).

Comment (TimSimpsonR) [inline .travis.yml]: Got it.

File: test-requirements.txt [hunk 11]
```diff
@@ -0,0 +1,3 @@
+pytest
+coverage
+mock
```
Comment (ehashman) [inline test-requirements.txt]: Are we using mock anywhere in this PR? I can't seem to find it.

**Comment by ehashman** [comment]
Modulo that last comment (and anything outstanding from @derwolfe), this LGTM.

File: tests/test_mutualauth.py (hunk 8 above)
Comment (derwolfe) [inline tests/test_mutualauth.py:38]: I'd be ok with that - then the test that has code using a real object returned by `_create_fake_resource` would be the only one declaring it. Further, it would allow `TestAuthorizeWithResource` to not need to inherit from this test. It could have its own `setUp` method where `self._resource` is set.

File: test-requirements.txt (hunk 11 above)
Comment (TimSimpsonR) [inline test-requirements.txt]: Oh thanks. I'd used it at one point but took it out.

**Comment by derwolfe** [comment]
LGTM once `mock` is removed.

**Comment by TimSimpsonR** [comment]
@derwolfe @ehashman Done.
